dev/x86_64/src/polyvec_basemul_acc_montgomery_cached_asm_k3.S

```diff
@@ -0,0 +1,160 @@
+ * SPDX-License-Identifier: Apache-2.0 OR ISC OR MIT
+ */
+
+/* References
+ * ==========
+ *
+ * - [REF_AVX2]
+ *   CRYSTALS-Kyber optimized AVX2 implementation
+ *   Bos, Ducas, Kiltz, Lepoint, Lyubashevsky, Schanck, Schwabe, Seiler, Stehlé
+ *   https://github.com/pq-crystals/kyber/tree/main/avx2
+ */
+
+/*
+ * This file is derived from the public domain
+ * AVX2 Kyber implementation @[REF_AVX2].
+ *
+ * The main difference is the use of a mulcache.
+ */
+
+#include "../../../common.h"
+#if defined(MLK_ARITH_BACKEND_X86_64_DEFAULT) && \
+    !defined(MLK_CONFIG_MULTILEVEL_NO_SHARED) && \
+    (defined(MLK_CONFIG_MULTILEVEL_WITH_SHARED) || MLKEM_K == 3)
+/* simpasm: header-end */
+
+#include "consts.h"
+
+#define _q %ymm0
+#define _qinv %ymm1
+#define _a %ymm2
+#define _b %ymm3
+#define _c %ymm4
+#define _d %ymm5
+#define _dz %ymm6
+#define _r0 %ymm7
+#define _r1 %ymm8
+#define _s0 %ymm9
+#define _s1 %ymm10
+#define _ac_hi %ymm11
+#define _ad_hi %ymm12
+#define _bdz_hi %ymm13
+#define _bc_hi %ymm14
+#define _a_lo %ymm13
+#define _b_lo %ymm14
+
+/* Polynomials to be multiplied are denoted (a + bX) (rsi arg) and
+ * (c + dX) (rdx arg). The mulcache for (c + dX), that is, the precomputed
+ * values of (d * zeta) is passed in rcx. We compute:
+ *  (r + sX) = (a + bX) * (c + dX), where in the code we denote
+ *    r = ac + bdz = r0 + r1,
+ *    s = ad + bc = s0 + s1.
+ */
+.macro schoolbook iter k
+vmovdqa (256*\k + 32*\iter +  0)*2(%rsi), _a
+vmovdqa (256*\k + 32*\iter + 16)*2(%rsi), _b
+vmovdqa (256*\k + 32*\iter +  0)*2(%rdx), _c
+vmovdqa (256*\k + 32*\iter + 16)*2(%rdx), _d
+vmovdqa (128*\k + 16*\iter)*2(%rcx), _dz
+
+/* Prepare Montgomery twists */
+vpmullw _a, _qinv, _a_lo
+vpmullw _b, _qinv, _b_lo
+
+/* Compute low-parts of monomials in (a + bX) * (c + dX),
+ * using Montgomery twists calculated before. */
+vpmullw _a_lo, _c,  _r0
+vpmullw _a_lo, _d,  _s0
+vpmullw _b_lo, _dz, _r1
+vpmullw _b_lo, _c,  _s1
+
+/* Compute the second high multiplication in Montgomery multiplication. */
+vpmulhw _r0, _q, _r0
+vpmulhw _s0, _q, _s0
+vpmulhw _r1, _q, _r1
+vpmulhw _s1, _q, _s1
+
+/* Compute high-parts of monomials in (a + bX) * (c + dX). */
+vpmulhw _a, _c,  _ac_hi
+vpmulhw _a, _d,  _ad_hi
+vpmulhw _b, _dz, _bdz_hi
+vpmulhw _b, _c,  _bc_hi
+
+/* Finish Montgomery multiplications */
+vpsubw _r0, _ac_hi, _r0
+vpsubw _s0, _ad_hi, _s0
+.if \iter & 1 /* Every other (d * zeta) is stored negative */
+vpsubw _bdz_hi, _r1, _r1
+.else
+vpsubw _r1, _bdz_hi, _r1
+.endif
+vpsubw _s1, _bc_hi, _s1
+
+vpaddw _r0, _r1, _r0
+vpaddw _s0, _s1, _s0
+.if \k > 0
+vmovdqa (32*\iter +  0)*2(%rdi), _r1
+vmovdqa (32*\iter + 16)*2(%rdi), _s1
+vpaddw _r0, _r1, _r0
+vpaddw _s0, _s1, _s0
+.endif
+
+vmovdqa _r0, (32*\iter +  0)*2(%rdi)
+vmovdqa _s0, (32*\iter + 16)*2(%rdi)
+/* Bounds: Since we are multiplying with signed canonical twiddles,
+ * each Montgomery multiplication has absolute value < q,
+ * and hence the coefficients of the output have absolute value < 2q. */
```

Comment:
Those comments need adjusting, I think. I don't think they were accurate before, but now we also need to take into account the accumulation.

As per the CBMC spec, what we know here is that `a`, `b` are < 4096 in absolute value, while we assume nothing about `c, d, dz` (e.g. 2^15 max absolute value). Then I think every individual Montgomery multiplication is bounded above by `3713` if I got the math right. And then we accumulate up to 8 of those. Luckily, `8 * 3713 = 29704 < 2^15`.